src/Dialect/ONNX/ONNXOps/Additional/FuncReturn.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===----------------- FuncReturn.cpp - ONNX Operations -------------------===//
+//
+// This file provides definition of ONNX dialect FuncReturn operation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+
+using namespace mlir;
+using namespace mlir::OpTrait::util;
+using namespace onnx_mlir;
+
+//===----------------------------------------------------------------------===//
+// Verify
+//===----------------------------------------------------------------------===//
+
+namespace {
+bool shapeIsSameOrMoreSpecific(ShapedType lhs, ShapedType rhs) {
+  // Unranked is the least specific:
+  if (!rhs.hasRank())
+    return true;
+  if (!lhs.hasRank())
+    return false;
+
+  // Otherwise the shapes are incompatible if ranks are different.
+  if (lhs.getRank() != rhs.getRank())
+    return false;
+
+  for (auto [lhsDimSize, rhsDimSize] :
+      llvm::zip(lhs.getShape(), rhs.getShape())) {
+    // Dim size is less specific or incompatible unless they are
+    // identical or rhs is dynamic ("?").
+    if (!ShapedType::isDynamic(rhsDimSize) && lhsDimSize != rhsDimSize)
```

Comment:
Should you also test if lhs is dynamic? It feels somewhat asymetrical. If desired, then there should be explanation in the function header that one arg is meant to be one type and the other the other type of param.

Reply:
added explanation of the asymmetry